### PR TITLE
Blog Post: Dataframe Protocol Support to Vaex

### PR DESCRIPTION
**Pull request by AlenkaF**
Hi! This blog post is another part of the internship series. The topic covered is the implementation of the dataframe interchange protocol for Vaex library. Hope you like it ❤️ 

## Text styling

- [x] The blog is written with [plain language](https://www.plainlanguage.gov/guidelines/) (where relevant).
- [x] If there are headers, they use the proper header tags in order to do so (with only one level-one header).
- [x] All links describe where they link to (for example, check the [Quansight labs website](https://labs.quansight.org/)).
- [x] Any kind of styling that the author uses (for example, bold for emphasis) is consistent throughout the blog.

## Non-text contents

- [x] All content is represented as text (for example, images need alt text and videos need captions or descriptive transcripts).
- [x] If there are emojis, there are not more than three in a row.
- [x] Don't use [flashing gifs or videos](https://www.w3.org/TR/UNDERSTANDING-WCAG20/seizure-does-not-violate.html).
- [ ] If it were to be read as plain text, the blog still makes sense and no information is missing.

### CODE REVIEW

**Review by rgommers — COMMENTED** [review]
Thanks Alenka! This reads really well and the illustrations are beautiful.

A small request about the images, in particular the one that's 1.5 MB - can you try running those through a filter to compress them a bit more? I'm not sure which one works best, but something like https://tinypng.com/.

(I haven't reviewed the text in detail)

**Comment by AlenkaF** [comment]
Thanks for pointing that out Ralf, I totally overlooked the pics size! Will do.

**Review by melissawm — COMMENTED** [review]
Nice work, @AlenkaF ! This is a tough post to write because of all the technical details, I think you found a good balance here, well done. I left a few suggestions but nothing major :) Cheers!

File: posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+**Implementation of the dataframe protocol** connects Vaex `dataframe` class with the `__dataframe__` method specified by the Consortium for Python Data API Standards. It also defines the constructor function called `from_dataframe ` with which the conversion between dataframes is made.
+
+The `__dataframe__` method returns an instance of the `_DataFrame` class. At this point I would like to quickly describe the conceptual model of how the data is represented in memory. After that I will go through the steps I took to implement the protocol for the Vaex library.
+
+> The smallest building blocks in the memory representation of the dataframe are 1-D arrays (or "buffers"), which are contiguous in memory and contain data with the same dtype. A column consists of one or more 1-D arrays (if, e.g., missing data is represented with a boolean mask, that's a separate array). A dataframe contains one or more columns. A column or a dataframe can be "chunked"; a chunk is a subset of a column or dataframe that contains a set of (neighboring) rows.
+
+<p align="center">
+    <img
+     alt="UML diagram of the classes `_Buffer`, `_Column` and `_DataFrame`."
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md:68]: Well, there is none as far as I am aware. It is my interpretation, which I hope is correct :)
I could insert a link to the [protocol blueprint](https://github.com/data-apis/dataframe-api/blob/main/protocol/dataframe_protocol.py)?

[the remark this answers is not in the document]
Comment (AlenkaF) [reply posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md:68]: Or maybe this is better: https://data-apis.org/dataframe-protocol/latest/API.html

File: posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+
+### Array interface and buffers
+
+When the `from_dataframe` method iterates through the columns it basically transfers the problem to columnar level where the interchange of data is really happening. That means array API is used.
+
+**Array API?** Just like we want to convert one type of dataframe into another, there are different protocols that do similar with columns/arrays/tensors. This is what is meant when array API term is used. For more information see https://data-apis.org/array-api/latest/design_topics/data_interchange.html.
+
+The best candidate for the array protocol is chosen to be [DLPack](https://github.com/dmlc/dlpack). Pandas and Vaex don't have this protocol implemented so instead we interpret a raw pointer through the use of [`array_interface`](https://numpy.org/devdocs/reference/arrays.interface.html). Therefore in Pandas and Vaex implementations the columns are read as NumPy arrays (ndarrays) via `array_interface`.
```
[the remark this answers is not in the document]
Comment (AlenkaF) [reply posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md]: Done. Hope it works?

File: posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md:67]: Me don't do UML diagrams.

The link to `API.html` should work.

File: posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (melissawm) [reply posts/2021/09/dataframe-interchange-protocol-and-the-vaex-library.md]: It does for me, thanks!

**Review by melissawm — APPROVED** [review]
Looks good to me, thanks @AlenkaF !

**Review by rgommers — APPROVED** [review]
This post is excellent, thanks again @AlenkaF! 

I pushed a commit with a few small tweaks and updated the publication date to today. There's one thing you may want to check: the Binder notebook has a first image that doesn't load for me, unsure why. All the code there does work fine.

**Comment by rgommers** [comment]
It's up at https://labs.quansight.org/blog/2021/10/dataframe-interchange-protocol-and-vaex/

**Pull request merged**